prometheus-connector/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusSinkWriter.java

```diff
@@ -0,0 +1,267 @@
+        if (LOG.isTraceEnabled()) {
+            LOG.trace(
+                    "Writing {} time-series containing {} samples ", timeSeriesCount, sampleCount);
+        }
+
+        Remote.WriteRequest writeRequest = buildWriteRequest(requestEntries);
+        byte[] requestBody;
+        try {
+            requestBody = compressWriteRequest(writeRequest);
+        } catch (IOException e) {
+            throw new RuntimeException("Exception compressing the request body", e);
+        }
+
+        SimpleHttpRequest postRequest = requestBuilder.buildHttpRequest(requestBody);
+        asyncHttpClient.execute(
+                postRequest,
+                new ResponseCallback(timeSeriesCount, sampleCount, counters, requestResult));
+    }
+
+    @VisibleForTesting
+    static class ResponseCallback implements FutureCallback<SimpleHttpResponse> {
+        private final int timeSeriesCount;
+        private final long sampleCount;
+        private final Consumer<List<Types.TimeSeries>> requestResult;
+        private final SinkCounters counters;
+
+        public ResponseCallback(
+                int timeSeriesCount,
+                long sampleCount,
+                SinkCounters counters,
+                Consumer<List<Types.TimeSeries>> requestResult) {
+            this.timeSeriesCount = timeSeriesCount;
+            this.sampleCount = sampleCount;
+            this.requestResult = requestResult;
+            this.counters = counters;
+        }
+
+        @Override
+        public void completed(SimpleHttpResponse response) {
+            if (RemoteWriteResponseClassifier.isSuccessResponse(response)) {
+                LOG.debug(
+                        "{},{} - successfully posted {} time-series, containing {} samples",
+                        response.getCode(),
+                        response.getReasonPhrase(),
+                        timeSeriesCount,
+                        sampleCount);
+                counters.inc(NUM_SAMPLES_OUT, sampleCount);
+                counters.inc(NUM_WRITE_REQUESTS_OUT);
+            } else {
+                counters.inc(NUM_SAMPLES_DROPPED, sampleCount);
+                counters.inc(NUM_WRITE_REQUESTS_PERMANENTLY_FAILED);
+
+                String responseBody = response.getBodyText();
+                if (RemoteWriteResponseClassifier.isNonRetriableErrorResponse(response)) {
+                    LOG.warn(
+                            "{},{} {} (discarded {} time-series, containing {} samples)",
+                            response.getCode(),
+                            response.getReasonPhrase(),
+                            responseBody,
+                            timeSeriesCount,
+                            sampleCount);
+                    counters.inc(NUM_SAMPLES_NON_RETRIABLE_DROPPED, sampleCount);
+                } else if (RemoteWriteResponseClassifier.isRetriableErrorResponse(response)) {
+                    LOG.warn(
```

Comment:
We shouldn't be dropping records here on a retryable error.

IMO we should either requeue the records, or propagate up the exception to fail the job.

Reply:
After offline discussion, we agreed that maybe we can set the HttpCllient to a lower maxRetryCount (by default), and then if the maxRetry is exceeded, we can throw the Exception to restart the Flink job